rslib/src/import_export/gather.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::{
+    collections::{HashMap, HashSet},
+    path::{Path, PathBuf},
+};
+
+use itertools::Itertools;
+
+use crate::{
+    card::{CardQueue, CardType},
+    decks::NormalDeck,
+    latex::extract_latex,
+    prelude::*,
+    revlog::RevlogEntry,
+    search::{Negated, SearchNode, SortMode},
+    storage::ids_to_string,
+    text::{
+        extract_media_refs, extract_underscored_css_imports, extract_underscored_references,
+        is_remote_filename,
+    },
+};
+
+#[derive(Debug, Default)]
+pub(super) struct ExportData {
+    pub(super) decks: Vec<Deck>,
+    pub(super) notes: Vec<Note>,
+    pub(super) cards: Vec<Card>,
+    pub(super) notetypes: Vec<Notetype>,
+    pub(super) revlog: Vec<RevlogEntry>,
+    pub(super) decks_configs: Vec<DeckConfig>,
+    pub(super) media_paths: HashSet<PathBuf>,
+}
+
+fn sibling_search(notes: &[Note], cards: &[Card]) -> SearchBuilder {
+    let mut nids = String::new();
+    ids_to_string(&mut nids, notes.iter().map(|note| note.id));
+    let mut cids = String::new();
+    ids_to_string(&mut cids, cards.iter().map(|card| card.id));
+    SearchBuilder::from(SearchNode::NoteIds(nids)).and(SearchNode::CardIds(cids).negated())
+}
+
+fn optional_deck_search(deck_id: Option<DeckId>) -> SearchNode {
+    if let Some(did) = deck_id {
+        SearchNode::from_deck_id(did, true)
+    } else {
+        SearchNode::WholeCollection
+    }
+}
+
+fn is_local_base_name(name: &str) -> bool {
+    !is_remote_filename(name) && Path::new(name).parent().is_none()
+}
+
+impl ExportData {
+    /*
+       pub(super) fn new(, media_folder: Option<PathBuf>) -> Self {
+           Self {
+               with_scheduling,
+               media_folder,
+               ..Default::default()
+           }
+       }
+    */
+    pub(super) fn gather_data(
+        &mut self,
+        col: &mut Collection,
+        deck_id: Option<DeckId>,
+        with_scheduling: bool,
+    ) -> Result<()> {
+        self.decks = col.gather_decks(deck_id)?;
+        let search = optional_deck_search(deck_id);
```

Comment:
Makes sense. Maybe `export_apkg()` can just accept a search? The backend code could expose a oneof of reasonable options (a deck id, all, note ids...), and pass on the according search node.

Reply:
Just remembered why I chose a deck-based approach: The user might want to export a top-level deck where all cards are in subdecks. Then we would also want to export the "empty" parent.
If we accepted a SearchNode, we could check if it's a single deck search, and export empty parent decks accordingly. Not the cleanest solution, but accepting both note ids and a deck id isn't pretty, either.

Reply:
Accepting a SearchNode sounds good.

Good point about needing the parents, but that should apply in the specific notes case as well - we must implicitly include all parent decks of the decks that have been used, as the deck tree code in the importing Anki will expect the parents to exist.

Reply:
97c9dd4 removes the feature to specify a deck, and move siblings outside that deck into it. Can we get away with that?
First of all, it won't affect the majority of users who don't have siblings in different decks. And for those who do, including cards outside the selected deck may not be the expected behaviour anyway.
We could show a warning if an export includes decks and cards outside the selected decks.

Reply:
I was mainly thinking of the shared decks case, where it's the users who import the deck who tend to encounter the problems/confusion, rather than the person exporting them (similar to our template checks). Showing an error on export if the user has elected to export a particular deck and cards are outside it may even be a better option than moving cards, as it avoids surprising behaviour, and is simpler than attempting to move the cards to a different deck. And for local exports, users could do them via the browse screen if they wanted to avoid the error.

Reply:
I assume then there'd be a separate routine to check for these things in advance? I think something like this would be required anyway if we want the export process to be more customisable. E.g. we could detect notetype conflicts in advance, and let the user decide on a resolution strategy (mapping or skipping).

Reply:
Wouldn't notetype conflict checking need to be done at import time? Maybe you have something else in mind than I'm imagining?

I'm not too fussed about the timing of the check, though it would be nice to at least do prior to all media being exported.

Reply:
Sorry, I got confused with import/export. For both cases, some routine for preparing/checking that is called before the import/export call by the frontend seems to make sense. Then we can be more naive in the core routines. In this case, we can keep just exporting all decks of matching cards.

Reply:
Yep, sounds good 👍